lyft_dataset_sdk/utils/export_kitti.py

```diff
@@ -0,0 +1,316 @@
+        sample_tokens = self._split_to_samples(split_logs)
+        if self.samples_count is not None:
+            sample_tokens = sample_tokens[:self.samples_count]
+
+        with parallel_backend('threading', n_jobs=self.parallel_n_jobs):
+            Parallel()(
+                delayed(self.process_token_to_kitti)(sample_token)
+                for sample_token
+                in tqdm(sample_tokens)
+            )
+
+    def process_token_to_kitti(self, sample_token: str) -> None:
+        # Get sample data.
+        sample = self.lyft_ds.get('sample', sample_token)
+        sample_annotation_tokens = sample['anns']
+
+        lidar_token = sample['data'][self.lidar_name]
+        sd_record_lid = self.lyft_ds.get('sample_data', lidar_token)
+        cs_record_lid = self.lyft_ds.get(
+            'calibrated_sensor', sd_record_lid['calibrated_sensor_token']
+        )
+        for cam_name in self.cams_to_see:
+            cam_front_token = sample['data'][cam_name]
+            if self.get_all_detections:
+                token_to_write = sample_token
+            else:
+                token_to_write = cam_front_token
+
+            # Retrieve sensor records.
+            sd_record_cam = self.lyft_ds.get('sample_data', cam_front_token)
+            cs_record_cam = self.lyft_ds.get(
+                'calibrated_sensor', sd_record_cam['calibrated_sensor_token'])
+            cam_height = sd_record_cam['height']
+            cam_width = sd_record_cam['width']
+            imsize = (cam_width, cam_height)
+
+            # Combine transformations and convert to KITTI format.
+            # Note: cam uses same conventions in KITTI and nuScenes.
+            lid_to_ego = transform_matrix(
+                cs_record_lid['translation'],
+                Quaternion(cs_record_lid['rotation']),
+                inverse=False
+            )
+            ego_to_cam = transform_matrix(
+                cs_record_cam['translation'],
+                Quaternion(cs_record_cam['rotation']),
+                inverse=True
+            )
+            velo_to_cam = np.dot(ego_to_cam, lid_to_ego)
+
+            # Convert from KITTI to nuScenes LIDAR coordinates, where we apply velo_to_cam.
+            velo_to_cam_kitti = np.dot(
+                velo_to_cam, self.kitti_to_nu_lidar.transformation_matrix)
+
+            # Currently not used.
+            imu_to_velo_kitti = np.zeros((3, 4))  # Dummy values.
+            r0_rect = Quaternion(axis=[1, 0, 0], angle=0)  # Dummy values.
+
+            # Projection matrix.
+            p_left_kitti = np.zeros((3, 4))
+            # Cameras are always rectified.
+            p_left_kitti[:3, :3] = cs_record_cam['camera_intrinsic']
+
+            # Create KITTI style transforms.
+            velo_to_cam_rot = velo_to_cam_kitti[:3, :3]
+            velo_to_cam_trans = velo_to_cam_kitti[:3, 3]
+
+            # Check that the rotation has the same format as in KITTI.
+            assert (velo_to_cam_trans[1:3] < 0).all()
```

Comment:
Hey @stalkermustang,
I've got a question: 
Why did you skip [this](https://github.com/nutonomy/nuscenes-devkit/blob/master/python-sdk/nuscenes/scripts/export_kitti.py#L152) assertion test?  I see that it fails when inserted here.

Reply:
as [i comment in kernel](https://www.kaggle.com/stalkermustang/converting-lyft-dataset-to-kitty-format#647435):

> I deleted this assert because I did not have ideas on how to fix it, nor did I understand what it affects. I tried to run the visualization and see what's wrong. Didn't see anything. Then, as @alfredocheng wrote, i run inference on PointRCNN and render predictions. Didn't see any problem - wrong rotated bboxes, missing annotations...all fine :/
So if you see a bottleneck in this assert or have an idea how to fix it or what affect is - pls write me :)

Reply:
It seems (from variable naming) like direction of transromfing axis (rotate matrix) from camera to car VELOcity. In Kitti we always have a front camera, but in lyft we has 6 cameras around car - and they directions isn't like KITTY's

Reply:
I run debugger on only front camera and see difference between left and right parts of assert. Two values from array are simmilar, but one is different: [-0.,  1., -0.] and [0, -1, 0]. If it's a "pi" - this is a similar rotation matrix.

Reply:
Okay, so velo means velodyne (name of the lidar sensor in KITTI), I've tested the original conversion script with nuscenes dataset and it worked well with the front camera and because lyft follows nuscenes data format this should work, I'll look into this.

Reply:
For front camera it's work well when we change -1 * pi to 1 * pi, but it's simmilar. Other arrays ([0, 0, -1], [1, 0, 0]) are the same and don't raise asser error. For another cameras this assert 
need not be True

For a long time I was unsure of the correctness of the converter, which is why I added a drawing results to visually check if there are any problems - in particular with bboxes from other cameras.

Reply:
>when we change -1 * pi to 1 * pi,
Could you please tell me where are you changing this value? 
here: `self.kitti_to_nu_lidar = Quaternion(axis=(0, 0, 1), angle=np.pi/2)`?

Reply:
> For other cameras this assert need not be True

makes sense.